source/shared/core_conn.cpp

```diff
@@ -150,34 +150,134 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
                 throw core::CoreException();
             }
-            if ( !is_missing_driver) {
-                break;
+            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION( odbc_version ) ];
+            r = core_odbc_connect( conn, conn_str_driver, is_pooled );
+        }  // else driver_version not unknown
+    } // else ce_option enabled
+#else
+    if( conn->ce_option.enabled ) {
+        // when AE is enabled, must use ODBC driver 17
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
+            CHECK_CUSTOM_ERROR( conn->driver_version != ODBC_DRIVER_17, conn, SQLSRV_ERROR_AE_DRIVER_REQUIRED, get_processor_arch() ) {
+                throw core::CoreException();
             }
-        } // for
+
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+
+            if( ! SQL_SUCCEEDED( r ) && core_compare_error_state( conn, r, "IM002" ) ) {
+                // sql state IM002 means that the specified ODBC driver is not installed
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND ) {
+                    throw core::CoreException();
+                }
+            }
+        }
+        else {
+            // driver not specified, so connect using ODBC 17
+            conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION::ODBC_DRIVER_17 ];
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+
+            if(! SQL_SUCCEEDED( r ) ) {
+                // sql state IM002 means that the specified ODBC driver is not installed
+                CHECK_CUSTOM_ERROR( core_compare_error_state( conn, r, "IM002" ) , conn, SQLSRV_ERROR_AE_DRIVER_REQUIRED, get_processor_arch() ) {
+                    throw core::CoreException();
+                }
+            }
+        } 
+    }
+    else { 
+        // column encryption NOT enabled
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+
+            if( ! SQL_SUCCEEDED( r ) && core_compare_error_state( conn, r, "IM002" ) ) {
+                // sql state IM002 means that the specified ODBC driver is not installed
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND, CONNECTION_STRING_DRIVER_NAME[ conn->driver_version ].c_str() ) {
+                    throw core::CoreException();
+                }
+            }
+        }
+        else {
+            for( short i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
+                std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION(i) ];
+                r = core_odbc_connect( conn, conn_str_driver, is_pooled );
+
+                if( SQL_SUCCEEDED( r ) ) {
+                    // successfully connected!
+                    break;
+                }
+                else if(! core_compare_error_state( conn, r, "IM002" ) ) {
```

Comment:
I did that before but this breaks one of the tests which specifically put in invalid option for a connection keyword

Reply:
hmm on 1st thought I wanted to say to fix the test...
but thinking more about this I think it's ok to stop there. While it's true that there can be other reasons for failing to connect to ODBC 13 I think it's fine to stop and not continue and try ODBC 11 because f you have ODBC 13 installed then I would think you do want to use ODBC 13. Why else would have have a newer version? If you wanted to use ODBC 11 you would've specified that explicitly. So if for some reason there was a failure connecting to ODBC 13 then I I think you would want to know and fix that instead of automatically going down to ODBC 11.